function_calling/src/tools.py

```diff
@@ -0,0 +1,163 @@
+        'X': operator.mul,
+        '÷': operator.truediv,
+        '/': operator.truediv
+    }
+    tokens = re.findall(r'\d+\.?\d*|\+|\-|\*|\/|÷|x|X', expression)
+    
+    if not tokens:
+        raise ToolException(f"Invalid expression '{expression}', should only contain one of the following operators + - * and /")
+    
+    current_value = float(tokens.pop(0))
+    
+    while tokens:
+        # The next token should be an operator
+        op = tokens.pop(0)
+        
+        # The next token should be a number
+        if not tokens:
+            raise ToolException(f"Incomplete expression '{expression}'")
+        try:
+            next_value = float(tokens.pop(0))
+            
+        except ValueError:
+            raise ToolException("Invalid number format")
+        
+        except:
+            raise ToolException("Invalid operation")
+        
+        # check division by 0
+        if op in ['/','÷'] and next_value == 0:
+            raise ToolException("cannot divide by 0")
+        
+        current_value = ops[op](current_value, next_value)
+    
+    return current_value
+
+
+# tool error handler
+def _handle_error(error: ToolException) -> str:
+    return f"The following errors occurred during Calculator tool execution: `{error.args}`"
+
+
+# tool definition
+calculator = StructuredTool.from_function(
+    func=calculator,
+    args_schema=CalculatorSchema,
+    handle_tool_error=_handle_error,#True,
+)
+
+## Python repl
+
+# tool schema
+class ReplSchema(BaseModel):
+    "A Python shell. Use this to execute python commands. Input should be a valid python commands and expressions. If you want to see the output of a value, you should print it out with `print(...), if you need a specific module you should import it`."
+    command: str = Field(..., description="python code to execute to evaluate")
+    
+# tool definition
+python_repl = PythonREPL()
+python_repl = Tool(
+    name="python_repl",
+    description="A Python shell. Use this to execute python commands. Input should be a valid python command. If you want to see the output of a value, you should print it out with `print(...)`.",
+    func=python_repl.run,
+    args_schema=ReplSchema
+)
+
+## SQL tool
+# tool schema
+class QueryDBSchema(BaseModel):
+    "A database querying tool. Use this to generate sql querys and retrieve the results from a database. Input should be a natural language question to the db."
+    query: str = Field(..., description="python code to execute to evaluate")
+    
+@tool(args_schema=QueryDBSchema)
+def query_db(query):
+    
+    """A database querying tool. Use this to generate sql querys and retrieve the results from a database. Input should be a natural language question to the db."""
+
+    fine_tunned_sql_llm = "sql_expert"
+    llm= SambaStudio(
+        model_kwargs={       
+                "max_tokens_to_generate": 2048,
+                "select_expert": fine_tunned_sql_llm,
+                "process_prompt": False
```

Comment:
Could you explain why the `process_prompt` should be False?

Reply:
removed parameter, pendign test sql with this new propmt with llama3